onnxruntime/core/providers/cuda/tensor/slice_impl.cu

```diff
@@ -8,95 +8,92 @@
-      int value = id;
       int dim = 0;
 #pragma unroll
       for (; dim < DIMS - 1; ++dim) {
-        output_strides[dim].divmod(value, div, mod);
+        output_strides[dim].divmod(mod, div, mod);
         input_index += (starts[dim] + div * steps[dim]) * input_strides[dim];
-        value = mod;
       }
       input_index += starts[dim] + mod * steps[dim];
-      input_indices[i] = input_index;
-      id += NumThreadsPerBlock;
+      if (is_grad) {
+        output_data[input_index] = values[i];
+      } else {
+        values[i] = input_data[input_index];
+      }
+      id += kNumThreadsPerBlock;
     }
   }
 
-  if (is_grad) {
-    id = start;
-#pragma unroll
-    for (int i = 0; i < NumElementsPerThread; i++) {
-      if (id < N) {
-        output_data[input_indices[i]] = input_data[id];
-        id += NumThreadsPerBlock;
-      }
-    }
-  } else {
+  if (!is_grad) {
     id = start;
 #pragma unroll
-    for (int i = 0; i < NumElementsPerThread; i++) {
+    for (int i = 0; i < kNumElementsPerThread; ++i) {
       if (id < N) {
-        output_data[id] = input_data[input_indices[i]];
-        id += NumThreadsPerBlock;
+        output_data[id] = values[i];
+        id += kNumThreadsPerBlock;
       }
     }
   }
 }
 
-Status SliceImpl(cudaStream_t stream,
-                 const size_t element_size,
-                 const int32_t dimension_count,
-                 const TArray<int64_t>& starts,
-                 const TArray<int64_t>& steps,
-                 const TArray<int64_t>& input_strides,
-                 const TArray<fast_divmod>& output_strides,
-                 const void* input_data,
-                 void* output_data,
-                 const size_t N) {
-  return SliceImplEx<false>(stream, element_size, dimension_count, starts, steps, input_strides, output_strides, input_data,
-                            output_data, N);
+Status SliceImpl(cudaStream_t stream, const size_t element_size, const int32_t dimension_count,
+                 const TArray<int64_t>& starts, const TArray<int64_t>& steps, const TArray<int64_t>& input_strides,
+                 const TArray<fast_divmod>& output_strides, const void* input_data, void* output_data, const size_t N) {
+  return SliceImplEx<false>(stream, element_size, dimension_count, starts, steps, input_strides, output_strides,
+                            input_data, output_data, N);
 }
 
-Status SliceImplGrad(cudaStream_t stream,
-                     const size_t element_size,
-                     const int32_t dimension_count,
-                     const TArray<int64_t>& starts,
-                     const TArray<int64_t>& steps,
-                     const TArray<int64_t>& input_strides,
-                     const TArray<fast_divmod>& output_strides,
-                     const void* input_data,
-                     void* output_data,
+#ifdef ENABLE_TRAINING
+Status SliceImplGrad(cudaStream_t stream, const size_t element_size, const int32_t dimension_count,
```

Comment:
Should this be moved under training_ops dir?

Reply:
There are two ways we can do:
- Put SliceImplGrad in a new file to training folder and include slice_impl.h, then we need to put SliceImplEx to the header file, but SliceImplEx is an internal function.
- Use ENABLE_TRAINING just as what we are doing now, so we don't need SliceImplEx in the header file.

I am thinking the latter one is better. Our gather_elements_impl.cu is also following this way. Because the .cu kernel implementation already contains the Grad logic, so basically only moving SliceImplGrad to training folder doesn't really separate the forward and backward implementation, but introduce trouble to code browsing as we need to add two new .h/.cc file which do nothing. And as mentioned above, SliceImplEx is an internal function, which is only called by SliceImpl and SliceImplGrad, it's not proper to put it in the header file (actually previous code had it in the header, I submitted a new PR to remove it).

Reply:
This may not work as is when we create a dedicated build for on device training. For now, this works... if there are other kernels (gather elements) which follow the same then I will find a way to include these kernels in on device training builds.   Long story short we can merge this PR and I will address this issue when I start to work on the on device training item.